### DISCUSSION TITLE
Using TAA causing panic when  entity is despawned

### DISCUSSION THREAD

**Discussion by PatrickChodowski** [bevyengine/bevy#9719] [Q&A]
Hello,

I added TAA to my game and I noticed it causes panic whenever I reload the scene or randomly when one of entities gets despawned:

```
thread 'main' panicked at 'error[B0003]: Could not insert a bundle (of type `bevy_pbr::prepass::PreviousGlobalTransform`) for entity 21v0 because it doesn't exist in this World.', /home/patrick/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_ecs-0.11.2/src/system/commands/mod.rs:957:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Encountered a panic when applying buffers for system `bevy_pbr::prepass::update_mesh_previous_global_transforms`!
Encountered a panic in system `bevy_app::main_schedule::Main::run_main`!
```

Is there a way to avoid it? Should I set the TemporalAntiAliasSettings::reset to true everytime I am about to despawn entities? Or maybe there is a way for TAA to not track certain entities? Thank you for help

**Comment by nicopap** [comment]
This is probably a bevy bug. I opened an issue for you

**Reply by PatrickChodowski** [reply]
Thank you!

**Reply by alice-i-cecile** [reply]
Said issue: https://github.com/bevyengine/bevy/issues/9721

**Comment by JMS55** [comment]
I worked on TAA, thanks for bringing this up! 

I can't figure out why this is happening. Are you deleting any entities in PreUpdate by chance?

**Reply by PatrickChodowski** [reply]
Hello, yes actually thats exactly the case! :)

Whenever I change  the scene I first despawn  all entities from previous scene in PreUpdate, and follow up with loading entities from new scene in Update

**Reply by PatrickChodowski** [reply]
For the record I wasnt aware that it is bad approach to have despawning in PreUpdate :) (if it is - in that case its not really a bug)

**Reply by nicopap** [reply]
No this is a bug, despawning entities in `PreUpdate` should not cause panics.

**Reply by JMS55** [reply]
What's the solution to this do you think? Stick an apply_deferred right after update_mesh_previous_global_transforms?

**Reply by nicopap** [reply]
Hue tough one.

I think this needs a different approach:

- The "insert `PreviousGlobalTransform` component" step should be separated from the "update `PreviousGlobalTransform` component" (supposedly in system with a `With<Handle<Mesh>>, Without<PreviousGlobalTransform>`)
- Now `update_mesh_previous_global_transforms` only mutates an existing component, no need to worry about command synchronization.

**Reply by JMS55** [reply]
How does that prevent the crash? The user could still delete the entity after 1., before commands are applied...

**Reply by nicopap** [reply]
Well, with this, the command only occurs the frame after the entity is spawned, rather than every frame. Which is not often the frame when the entity is despawned… Not optimal, maybe feels like dusting the crap under the rug wrt this issue specifically. But in terms of performance, it's definitively better, since you are only spending time in exclusive-world-access world when the component is added, not every frame for each mesh entity with a global transform.

To fix the final issue, there should be a non-failing insert command. The original `Insert` is defined there:

https://github.com/bevyengine/bevy/blob/17edf4f7c7a6caf0862322286f27d3e15ef0f5c1/crates/bevy_ecs/src/system/commands/mod.rs#L942-L960

**Reply by alice-i-cecile** [reply]
A `try_insert` command was added in https://github.com/bevyengine/bevy/pull/9844 and will be releasing with Bevy 0.12.
